### PR TITLE
 [ADD] mrp_operations_schedule

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
This module was written to extend the manufacturing capabilities of Odoo,
and adds automatic re-scheduling capabilities to the manufacturing order
and the associated work orders. It also shows a gantt chart view that
adjusts automatically during the execution of the manufacturing order and
work orders.

### CODE REVIEW

File: mrp_operations_schedule/model/mrp.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import fields, osv, orm
```
Comment (pedrobaeza) [inline mrp_operations_schedule/model/mrp.py:22]: Use new API if possible, but at least orm.Model

File: mrp_operations_schedule/model/mrp.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+from datetime import datetime
+
+
+class mrp_production(osv.osv):
```
Comment (pedrobaeza) [inline mrp_operations_schedule/model/mrp.py]: CamelCase

File: mrp_operations_schedule/model/mrp.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        return False
+
+    def _reschedule_workorders(self, cr, uid, ids, context=None):
+        """ Reschedules the work orders"""
+        dt_end = datetime.now()
+        if context is None:
+            context = {}
+        for po in self.browse(cr, uid, ids, context=context):
+            dt = dt_end = datetime.strptime(po.date_planned,
+                                         '%Y-%m-%d %H:%M:%S')
```
Comment (pedrobaeza) [inline mrp_operations_schedule/model/mrp.py]: Incorrect indentation

File: mrp_operations_schedule/model/mrp.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+            if production.date_finished:
+                res[production.id]['date_end_gantt'] = \
+                    production.date_finished
+
+            start_dates = []
+            end_dates = []
+            for workorder in production.workcenter_lines:
+                # Add the actual start date if found. Otherwise the planned
+                # start date.
+                date_start = workorder.date_start or workorder.date_planned \
```
Comment (pedrobaeza) [inline mrp_operations_schedule/model/mrp.py]: Better to enclose multi-lines with ():

```
date_start = (workorder.date_start or workorder.date_planned
              or False)
```

File: mrp_operations_schedule/model/mrp.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                                          type='datetime',
+                                          multi='gantt_dates'),
+    }
+
+    def action_compute(self, cr, uid, ids, properties=None, context=None):
+        """ Computes bills of material of a product and planned date of work order.
+        @param properties: List containing dictionaries of properties.
+        @return: No. of products.
+        """
+        result = super(mrp_production, self).action_compute(cr, uid, ids, properties=properties, context=context)
```
Comment (pedrobaeza) [inline mrp_operations_schedule/model/mrp.py]: Don't put more than 80 cols

File: mrp_operations_schedule/model/mrp.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        @return: Dictionary of values.
+        """
+        ops = self.browse(cr, uid, ids, context=context)
+        date_and_hours_by_cal = []
+        for op in ops:
+            date_start = op.date_start or op.date_planned
+            if date_start:
+                date_and_hours_by_cal.extend(
+                    [(date_start, op.hour, op.workcenter_id.calendar_id.id)])
+        intervals = self.pool.get('resource.calendar').interval_get_multi(
```
Comment (pedrobaeza) [inline mrp_operations_schedule/model/mrp.py]: self.pool['resource.calendar']

File: mrp_operations_schedule/static/description/index.html [hunk 7]
```diff
@@ -0,0 +1,83 @@
+<section class="oe_container">
```
Comment (legalsylvain) [inline mrp_operations_schedule/static/description/index.html]: why is this file necessary ? it seems to contain the same information as readme.rst.
Regards.

**Comment by JordiBForgeFlow** [comment]
Fixed as per your feedback. Thanks @pedrobaeza and @legalsylvain

File: mrp_operations_schedule/README.rst [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+Usage
+=====
+
+No specific usage instructions are required.
+
+Known issues / Roadmap
+======================
+
+No issues have been identified.
+
```
Comment (clonedagain) [inline mrp_operations_schedule/README.rst]: Maybe you should drop the empty sections?

**Comment by JordiBForgeFlow** [comment]
Removed empty sections in README.rst as @clonedagain suggested.

File: mrp_operations_schedule/model/mrp.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+    def _compute_planned_workcenter(self, cr, uid, ids, context=None, mini=False):
+        return False
+
+    def _reschedule_workorders(self, cr, uid, ids, context=None):
+        """ Reschedules the work orders"""
+        dt_end = datetime.now()
+        if context is None:
+            context = {}
+        for po in self.browse(cr, uid, ids, context=context):
```
Comment (bealdav) [inline mrp_operations_schedule/model/mrp.py:38]: po is purchased_order ? or else ?

Comment (pedrobaeza) [inline mrp_operations_schedule/model/mrp.py:38]: I supposed "production order", but better to use usual mo abreviature

File: mrp_operations_schedule/model/mrp.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+                                            multi='gantt_dates'),
+        'date_end_gantt': fields.function(_get_gantt_dates,
+                                          string='Gantt end date',
+                                          type='datetime',
+                                          multi='gantt_dates'),
+        'date_planned_end': fields.function(_get_date_end,
+                                            string='End Date',
+                                            type='datetime',
+                                            store=False),
+    }
```
Comment (jco-odoo) [inline mrp_operations_schedule/model/mrp.py:199]: What is the purpose of these gantt dates?  Avoiding changing dates when moving in the calendar/gantt?

Comment (JordiBForgeFlow) [inline mrp_operations_schedule/model/mrp.py:199]: @jco-odoo Yes. IMHO you should respect the plan vs. actual. When was it planned to start and when did it actually start. When it was scheduled to end vs. when it was ended. Otherwise you will later totally loose the ability to review the performance (do we complete the MO's on time, or we are are always delayed)?

Comment (jco-odoo) [inline mrp_operations_schedule/model/mrp.py:199]: I would make date_planned_end a normal field and rename date_planned to
date_planned_start and use that.  There is a date_start and date_end
already.  What do you think?

Comment (JordiBForgeFlow) [inline mrp_operations_schedule/model/mrp.py:199]: Considering that in the MRP Workshop we talked about a Scheduler being in charge of re-scheduling Work Orders, that would make a lot of sense. The Scheduler changes the date_planned_start and date_planned_end. date_start and date_end are set when the user starts / ends the WO.

**Comment by pedrobaeza** [comment]
Is this going to be continued, @jbeficent ?

**Comment by pedrobaeza** [comment]
Closing as no answer. Feel free to reopen if needed.
